File: Long_Word.java
/*Running Time Complexity:O(nlogk)
Space Complexity: O(n)
Successfully Run and Compiled on leetcode
*/
class Solution {
    class TrieNode{
        TrieNode[] children;
        String word;
        public TrieNode(){
            this.children = new TrieNode[26];
        }
    }
    TrieNode root;
    public String longestWord(String[] words) {
        if(words==null) return "";
        root = new TrieNode();
        for(String word : words){
            insert(word);
        }
        Queue<TrieNode> q = new LinkedList<>();
        q.add(root);
        TrieNode curr=null;
        while(!q.isEmpty()){
            curr = q.poll();
            for(int i = 25;i>=0;i--){
                if(curr.children[i]!=null && curr.children[i].word != null){
                    q.add(curr.children[i]);
                }
            }
        }
       if(curr.word==null)return ""; 
       return curr.word; 
    }
    private void insert(String words){
        
        TrieNode curr = root;
        for(int i =0;i<words.length();i++){
            char c = words.charAt(i);
            if(curr.children[c-'a']==null){
                curr.children[c-'a'] = new TrieNode();
            }
            curr = curr.children[c-'a'];
        }
        curr.word = words;
        
    }
}

File: LongWord.py
# BFS
# Time Complexity  nlogn + n *(2* avg len of word)
# Space Complexity o(n) sum of length all of the words
class TrieNode:
    
    def __init__(self,char):
        
        self.children = [None]*26
        self.isEnd = True
        self.char = char
        self.isWord = False
        self.word = None
            

class Solution:
    
    def __init__(self):
        self.trie = TrieNode("0")
        self.trie.isWord == True
    
    def insert(self,word):
        
        curr = self.trie
   
        for a in word:
            
            childIndex = ord(a) - ord('a')
            
            if curr.children[childIndex] == None:
                curr.children[childIndex] = TrieNode(a)
                
            curr = curr.children[childIndex]
        curr.isWord = True
        curr.word = word
    
            
           
            
    def longestWord(self, words: List[str]) -> str:
        
        for a in words:
            self.insert(a)
        qu = collections.deque()
        qu.append(self.trie)
        curr = 0
        while(qu):
            curr = qu.popleft()
            for i in range(len(curr.children)-1,-1,-1):
                
                if curr.children[i] != None and curr.children[i].isWord:
                        qu.append(curr.children[i])
                       
        
        if curr.word == None:
            return ""
            
        return curr.word
        
            
        
        
        
        
        
        
        